### PR TITLE
Add: implement the GLM-5.3-Flash KDA kernels for a2a3

### PR DESCRIPTION
**Pull request by lwDavid**
Implements the five KDA operators the scaffold left as stubs, so the
34 linear-attention layers of GLM-5.3-Flash run end to end on a2a3:
the front-end projections and gates, the short causal convolution,
the single-token recurrence, the chunked delta rule, and the gated
output norm with its row-parallel projection.

Every kernel is validated against a Torch reference on device, not
only in the simulator. That distinction earned its keep: three of the
failures below compile and pass under `a2a3sim` and fail on silicon.

## Numerics

The gate is `-5.0 * sigmoid(exp(A_log) * (f_b(f_a(x)) + dt_bias))` in
FP32 throughout. `A_log` is per head and broadcasts across all 128
channels while `dt_bias` is per channel; both flatten to the same 512
values per rank, so swapping them is invisible to every shape check.
The official reference and vLLM-Ascend agree here, and the softplus
branch beside it is unreachable at `gate_lower_bound = -5.0`, so it is
not implemented.

The recurrent state is `[rows, heads, V, K]`, the layout the AscendC
operator's ABI fixes. Both axes are 128, so a transposed state passes
every check and shows up only as wrong output from the second chunk
on. The scan therefore runs in that orientation and lets the matmul
operand flags do the re-orientation.

`decay` carries the **log** decay. The projection does not
exponentiate it; both delta-rule kernels do.

## Chunk length and the decay pivot

`CHUNK` is 16, against 64 in the reference and 128 in the Ascend-tuned
pypto implementation, and it is pinned by two constraints rather than
chosen. At 64 the chunk body holds eight `[64, 128]` FP32
intermediates and overruns UB. And the decay has to be reconstructed
around a pivot: `exp(gcum_i - gcum_j)` for a column ahead of the row
overflows to `+inf`, the mask then computes `inf * 0 = NaN`, and the
clamp that prevents it also truncates the *kept* lower-triangular
entries once the decay between pivot and column exceeds it. The gate
is bounded by `|g| <= 5` per token, so a 16-token span bounds that
accumulation at 75 and the clamp provably never reaches a kept entry.
Measured against a per-token reference in FP64: one pivot per 32-row
chunk is wrong by 4e-2, one per 16 rows is exact to 6e-17. Raising the
chunk again means several pivots per chunk and stitching their bands
together, which is the first thing to revisit when this is profiled.

## What the platform dictated

Three shapes in here look roundabout and are not:

* The triangular inverse is a four-step series, `E_0 = I - A` and
  `P_{j+1} = E_j P_j`, each step its own kernel chained by an explicit
  dependency. Row-at-a-time forward substitution is the textbook form,
  but its matmul has `M = 1`, which the cube lowers to an extract
  whose row count must be a multiple of 16. Within one kernel a store
  and a later read of the same buffer are not ordered, so the seed
  read back as zeros and collapsed the product; across kernels the
  dependency is what orders them.
* The inverse is carried at 128 columns with the 16 real ones on the
  left. A 16-wide FP32 tile has no hand-off between the cube and the
  vector pipe here, so the arithmetic has nowhere to run at the
  matrices' natural size. The padding stays zero through both
  recurrences.
* Chunk tails are zeroed with `fillpad`. `valid_shape` marks which
  rows are real; it does not zero the rest, and the decay's cumulative
  sum depends on the padding being zero. Sequences that are a whole
  number of chunks never expose this, which is why the case list
  leads with a ragged one.

## Corrections to the scaffold ABI

* `conv_state` becomes time-major `[rows, KDA_CONV_K - 1, CONV_DIM]`
  and `weight` becomes `[KDA_CONV_K, CONV_DIM]`, both following the
  AscendC kernel's own `[cache, state_len, dim]` convention. The
  channel-major forms need a transpose per request to produce rows the
  convolution already wants.
* `beta` carries the cube's 16-column padding. It is one value per
  head, four at TP16, and narrowing a wider tile down to four FP32
  columns does not work in any of its spellings: the store writes the
  tile's full row and spills each token's gate into the next three.
* `query_start_loc` uses `Q_START_DYN` rather than `B_DYN + 1`.
  Arithmetic on a dynamic symbol is not valid in a shape annotation.
* `config.py` gains `KDA_STATE_DYN` for the state pools. Reusing
  `B_DYN` asserts that pool depth equals batch size, which is exactly
  what `state_rows` exists to avoid.

## Fixtures

The keys share a per-head direction. Independent random keys are
near-orthogonal at width 128, which leaves the Gram matrix around 0.09
and the triangular inverse indistinguishable from the identity —
replacing the inverse with `I` outright still passed the first version
of this fixture. The state pools are non-zero and `state_rows` is not
the identity mapping, so a kernel that ignores the indirection fails.

Verified on device across 28 cases covering ragged and aligned token
counts, single-token batches, multi-request batches, and the two
isolation cases that remove the carried state and the inverse from the
path. MTP and performance tuning are out of scope here.

### CODE REVIEW

File: models/glm5_3_flash/kda_conv.py [hunk 1]
```diff
@@ -47,50 +93,390 @@ def golden_kda_conv_prefill(
...
+    with pl.spmd(requests * (CONV_DIM // C_TILE), name_hint="kda_conv_head",
+                 deps=[bulk_tid]):
+        task = pl.tile.get_block_idx()
+        lanes = CONV_DIM // C_TILE
+        b = task // lanes
+        c0 = (task % lanes) * C_TILE
+        s = pl.cast(pl.read(query_start_loc, [b]), pl.INDEX)
+        e = pl.cast(pl.read(query_start_loc, [b + 1]), pl.INDEX)
+        row = pl.cast(pl.read(state_rows, [b]), pl.INDEX)
+        length = e - s
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/glm5_3_flash/kda_conv.py:210]: Confirmed and fixed. The per-request body is now guarded by `if length > 0:`, so an empty request leaves `conv_state` untouched, and the ragged fixture carries a zero-length request (`[5, 1, 9, 2, 0]`) so the path is exercised.

The guard had one consequence worth recording: once `length` is known positive, `STATE_LEN - length - i` is provably non-positive at the last state row, so its old-state branch folds to a zero-row view and the compiler rejects it outright (`partition_view op size at dim 0 must be positive, got 0`). That row is always fed by the stream under the same premise, so it is now written directly instead of through the masked pair.

All five `kda_conv` cases pass on device.

File: models/glm5_3_flash/kda_output.py [hunk 2]
```diff
@@ -45,7 +76,142 @@ def kda_output(
...
+        gated_rows[r0 : r0 + ROW_TILE, 0:KDA_DIM] = pl.cast(activated, pl.BF16, mode="rint")
+
+    # The padded tail rows are never written, so zero them: the projection reads whole
+    # tiles and a recycled scratch row would otherwise reach the cube as garbage.
+    if t_pad * LOCAL_KDA_H > t_dim * LOCAL_KDA_H:
+        with pl.spmd(1, name_hint="kda_output_pad"):
+            pad0 = pl.tile.get_block_idx() * ROW_TILE + t_dim * LOCAL_KDA_H
+            gated_rows[pad0 : pad0 + ROW_TILE, 0:KDA_DIM] = pl.full(
+                [ROW_TILE, KDA_DIM], dtype=pl.BF16, value=0.0
+            )
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/glm5_3_flash/kda_output.py]: Confirmed and removed, not replaced. The norm pass runs one task per token tile over the padded count and each task stores a full `ROW_TILE`, so the whole allocation is written, and the rows past `t_dim` load through `valid_shape`, which zero-fills them. A comment now says so, to keep the next reader from adding the store back.

`kda_output` ragged/aligned/single pass on device.

File: models/glm5_3_flash/kda_projection.py [hunk 3]
```diff
@@ -64,14 +141,238 @@ def kda_projection(
...
+    # columns further along than they should. Materialising a zero-padded weight makes
+    # all 16 accumulator columns real, and the store is width-matched.
+    w_b_pad = pl.create_tensor([BETA_PAD, D], dtype=pl.BF16)
+    with pl.spmd(BETA_PAD, name_hint="kda_projection_b_pad") as b_pad_tid:
+        r = pl.tile.get_block_idx()
+        w_b_pad[r : r + 1, 0:D] = pl.slice(w_b, [1, D], [pl.min(r, LOCAL_KDA_H - 1), 0])
+    with pl.spmd(BETA_PAD - LOCAL_KDA_H, name_hint="kda_projection_b_zero",
+                 deps=[b_pad_tid]) as b_zero_tid:
+        r = LOCAL_KDA_H + pl.tile.get_block_idx()
+        w_b_pad[r : r + 1, 0:D] = pl.full([1, D], dtype=pl.BF16, value=0.0)
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/glm5_3_flash/kda_projection.py]: Real issue, but the suggested diff does not survive tracing. `b_zero_tid = b_pad_tid` aliases a task handle in a traced body, which the parser rejects, and folding the count inside the body fails the same way (`Unsupported function call: max(16 - 4, 1)`) — the body admits only `pl.*` calls.

What works is to fold at module scope and keep the region unconditional:

```python
B_ZERO_TASKS = max(BETA_PAD - LOCAL_KDA_H, 1)
...
with pl.spmd(B_ZERO_TASKS, name_hint="kda_projection_b_zero", deps=[b_pad_tid]) as b_zero_tid:
    r = LOCAL_KDA_H + pl.tile.get_block_idx()
    if r < BETA_PAD:
        w_b_pad[r : r + 1, 0:D] = pl.full([1, D], dtype=pl.BF16, value=0.0)
```

At TP4 that is one task whose body writes nothing, and `b_zero_tid` always exists for the main region's dependency.

File: models/glm5_3_flash/kda_projection.py [hunk 4]
```diff
@@ -64,14 +141,238 @@ def kda_projection(
...
+        TensorSpec("w_f_a", [KDA_DIM, D], bf, init_value=w(KDA_DIM, D)),
+        TensorSpec("w_f_b", [LOCAL_KDA_QKV_DIM, KDA_DIM], bf,
+                   init_value=w(LOCAL_KDA_QKV_DIM, KDA_DIM, 0.08)),
+        # dt_bias spans a wide range in the real checkpoint; a narrow fixture would
+        # hide a per-head / per-channel mix-up behind the sigmoid's flat tails.
+        TensorSpec("dt_bias", [LOCAL_KDA_QKV_DIM], f32,
+                   init_value=lambda: torch.randn(LOCAL_KDA_QKV_DIM) * 2.0),
+        # Distinct per-head values, so a head-axis broadcast error is visible.
+        TensorSpec("a_log", [LOCAL_KDA_H], f32,
+                   init_value=lambda: torch.tensor([-0.7, 0.0, 0.4, 1.1])[:LOCAL_KDA_H]),
```
[the remark this answers is not in the document]
Comment (lwDavid) [reply models/glm5_3_flash/kda_projection.py]: Confirmed and fixed: `torch.linspace(-0.7, 1.1, LOCAL_KDA_H)`. The old literal happened to match `LOCAL_KDA_H` at TP16 and would have failed the golden reshape anywhere else.

Distinct per-head rates are also the stronger fixture: one shared decay makes a dropped or mis-broadcast per-head `A_log` invisible, which is the failure this operator is most prone to, since `A_log` is per head and `dt_bias` is per channel and the two flatten to the same value count per rank.
